Source/ProgressionSystemRuntime/Private/Data/PSWorldSubsystem.cpp

```diff
@@ -37,3 +45,113 @@ void UPSWorldSubsystem::SetProgressionSystemComponent(UPSHUDComponent* MyProgres
+	// requires to be executed to guarntee that data asset will be loaded from the ini file
+	// otherwise the data asset file will not be loaded at fresh (heaviest) load of UE
+	PSDataAssetInternal = GetPSDataAsset();
+	if (PSDataAssetInternal)
+	{
+		ProgressionDataTableInternal = PSDataAssetInternal->GetProgressionDataTable();
+		LoadGameFromSave();	
+	} else
+	{
+		UE_LOG(LogProgressionSystem, Warning, TEXT("PSDataAssetInternal null"))
+	}
+}
+
+void UPSWorldSubsystem::OnCharacterPossessed(APawn* MyPawn)
+{
+	if (APlayerCharacter* MyPlayerCharacter = Cast<APlayerCharacter>(MyPawn))
+	{
+		MyPlayerCharacter->OnPlayerTypeChanged.AddUniqueDynamic(this, &ThisClass::OnPlayerTypeChanged);
+
+		//Unsubscribe to ignore null events call
+		AMyPlayerController* MyPC = UMyBlueprintFunctionLibrary::GetLocalPlayerController();
+		MyPC->GetOnNewPawnNotifier().RemoveAll(this);
+	}
+}
+
+void UPSWorldSubsystem::OnPlayerTypeChanged(FPlayerTag PlayerTag)
+{
+	if (SaveGameInstanceInternal)
+	{
+		for (const auto& KeyValue : SaveGameInstanceInternal->SavedProgressionRows)
+		{
+			FName Key = KeyValue.Key;
+			FPSRowData RowData = KeyValue.Value;
+
+			if (RowData.Map == UMyBlueprintFunctionLibrary::GetLevelType() && RowData.Character == PlayerTag)
+			{
+				SavedProgressionRowDataInternal = SaveGameInstanceInternal->SavedProgressionRows[Key];
+			}
+		}
+		SaveDataAsync();
+	}
+	OnCurrentRowDataChanged.Broadcast(SavedProgressionRowDataInternal);
+}
+
+void UPSWorldSubsystem::LoadGameFromSave()
+{
+	// Check if the save game file exists
+	if (UGameplayStatics::DoesSaveGameExist(SaveGameInstanceInternal->GetSaveSlotName(), SaveGameInstanceInternal->GetSaveSlotIndex()))
+	{
+		SaveGameInstanceInternal = Cast<UPSSaveGameData>(UGameplayStatics::LoadGameFromSlot(SaveGameInstanceInternal->GetSaveSlotName(), SaveGameInstanceInternal->GetSaveSlotIndex()));
+	}
+	else
+	{
+		// Save file does not exist
+		// do initial load from data table
+		TMap<FName, FPSRowData> SavedProgressionRows;
+		UMyDataTable::GetRows(*ProgressionDataTableInternal, SavedProgressionRows);
+		SaveGameInstanceInternal = Cast<UPSSaveGameData>(UGameplayStatics::CreateSaveGameObject(UPSSaveGameData::StaticClass()));
+
+		if (SaveGameInstanceInternal)
+		{
+			SaveGameInstanceInternal->SavedProgressionRows = SavedProgressionRows;
+		}
+	}
+	SetFirstElemetAsCurrent();
+}
+
+void UPSWorldSubsystem::SetFirstElemetAsCurrent()
+{
+	if (SaveGameInstanceInternal)
+	{
+		// Create an iterator for the TMap
+		TMap<FName, FPSRowData>::TIterator Iterator = SaveGameInstanceInternal->SavedProgressionRows.CreateIterator();
+
+		// Check if the iterator is pointing to a valid position
+		if (Iterator)
+		{
+			// Access the key and value of the first element
+			SavedProgressionRowDataInternal = Iterator.Value();
+			SavedProgressionRowDataInternal.IsLevelLocked = false;
+			Iterator.Value().IsLevelLocked = false;
+		}		
+		else
+		{
+			// The TMap is empty
+			UE_LOG(LogTemp, Warning, TEXT("The TMap is empty."));
```

Comment:
Please prefer ensureMsgf instead of logs as it does log+assert in code, but not just log.
````
aif (ensureMsgf(Iterator, TEXT("Assert: text"))
{
... do rest
}
```

Reply:
done